R/generate_data.R

```diff
@@ -309,30 +310,64 @@ genFormula <- function(coefs, vars) {
   lvars <- length(vars)
 
   if (!(lcoef == lvars | lcoef == lvars + 1)) {
-    stop("Coefficients or variables not properly specified")
-  }
-
-  if (!is.numeric(coefs)) {
-    stop("Coefficients must be specified as numeric values or numeric variables")
+    c <- condition(c("simstudy::coeffVar", "error"),
+                   "Coefficients or variables not properly specified!")
+    stop(c)
   }
 
-  if (!is.character(vars)) {
-    stop("Variable names must be specified as characters or character variables")
+  # check coefs are numeric or character
+  
+  # if (any(!is.numeric(coefs) | !is.character(coefs))) {
+  #   c <- condition(c("simstudy::coefNumOrChar", "error"),
+  #                  "Coefficients must be of type numeric or character!")
+  #   stop(c)
+  # }
+  
+  #assertNumeric(var1 = coefs)
+  # if (!is.numeric(coefs)) {
+  #   stop("Coefficients must be specified as numeric values or numeric variables")
+  # }
+
+  assertType(var1 = vars, type = "character")
+  # if (!is.character(vars)) {
+  #   stop("Variable names must be specified as characters or character variables")
+  # }
+  
+  double_dot_vars <- .parseDotVars(coefs)
+  
+  coef_gen <- function (cf) {
+    cf_char <- as.character(cf)
+    cff <- if(!is.null(double_dot_vars[[cf_char]])) {
+      double_dot_vars[[cf_char]]
+    } else {
+      cf_char
+    }
+    
+    # cff <- tryCatch(as.numeric(cf),
+    #                 warning = function (w) {
+    #                   paste0("..", cf)
+    #                 })
+                    
+    return(paste0(cff))
   }
 
   if (lcoef != lvars) { # Intercept
 
-    form <- paste0(coefs[1])
+    #form <- paste0(coefs[1])
+    form <- coef_gen(coefs[1])
     coefs <- coefs[-1]
   } else { # no intercept
 
-    form <- paste(coefs[1], "*", vars[1])
+    #form <- paste(coefs[1], "*", vars[1])
+    form <- paste(coef_gen(coefs[1]), "*", vars[1])
     coefs <- coefs[-1]
     vars <- vars[-1]
   }
 
   for (i in 1:(lcoef - 1)) {
-    form <- paste(form, "+", coefs[i], "*", vars[i])
+    #form <- paste(form, "+", coefs[i], "*", vars[i])
+    cg <- coef_gen(coefs[i])
+    form <- paste(form, "+", cg, "*", vars[i])
   }
 
   return(form)
```

Comment:
This does seem to work well - nice job - but I had another thought (even before looking at the testing). What if we used `genFormula` to be more literal and just generate a formula literally, so that

```
genFormula(c(1, "..x", "..y"), c("a", "b"))
```

just returns `"1 + ..x * a + ..y * b"` and doesn't do any parsing at all. When the formula is applied later on in `defData`, then the the proper checks will be conducted to make sure that the formula is valid. The only thing `genFormula` would care about is whether there are two vectors passed, and that they are either the same length (no intercept) or the first one is one element longer (the intercept). Either one could be any combo of any kind of strings (double dot or not) and numbers. So even something like this would be allowed:

```
genFormula(c(1, "x", "..y"), c("..a", "3"))
```

which would return `"1 + x * ..a + ..y * 3"` or even

```
genFormula(c(1, "..x", "5"), c("..a", "3"))
```

which would return `"1 + ..x *..a + 5 * 3"`

These could all be valid formulas in some data generating process, though the last one is equivalent to  `"16 + ..x *..a"`

I know you spent a lot of time figuring out how to parse the data, but what do you think about this simplification? And maybe it will make the testing easier?

Reply:
That does make sense to me, it would make the process more efficient for sure between the two functions. This also seems a bit more flexible because you can create a formula with already having the variables defined which could be good in some scenarios. I'll change it to that, getting it to work wasn't too bad, and it was a good exercise, so I really don't mind changing it. And yes, I think it would make the testing easier.